net/dns/resolver/forwarder.go

```diff
@@ -220,9 +284,14 @@ func (f *forwarder) packetListener(ip netaddr.IP) (packetListener, error) {
 // iOS and we want the number of pending DNS lookups to be bursty
 // without too much associated goroutine/memory cost.
 func (f *forwarder) send(ctx context.Context, txidOut txid, closeOnCtxDone *closePool, packet []byte, dst netaddr.IPPort) ([]byte, error) {
-	// TODO(bradfitz): if dst.IP is 8.8.8.8 or 8.8.4.4 or 1.1.1.1, etc, or
-	// something dynamically probed earlier to support DoH or DoT,
-	// do that here instead.
+	// Upgrade known DNS IPs to DoH (DNS-over-HTTPs).
+	if urlBase, dc, ok := f.getDoHClient(dst.IP()); ok {
+		res, err := f.sendDoH(ctx, urlBase, dc, packet)
```

Comment:
Before sending to DoH, the ID should be replaced with zero; the response should probably have the correct ID substituted back in.

> In order to maximize HTTP cache friendliness, DoH clients using media
> formats that include the ID field from the DNS message header, such
>   as "application/dns-message", SHOULD use a DNS ID of 0 in every DNS
>   request.  HTTP correlates the request and response, thus eliminating
>   the need for the ID in a media type such as "application/dns-
>   message".  The use of a varying DNS ID can cause semantically
>   equivalent DNS queries to be cached separately.

Reply:
I'm going to punt on this for now, as it's just a SHOULD and I'd rather not parse & re-pack the DNS queries yet.

I tested that it doesn't matter with this, though:

```diff
diff --git a/net/dns/resolver/doh_test.go b/net/dns/resolver/doh_test.go
index 978d76f1..e283f32e 100644
--- a/net/dns/resolver/doh_test.go
+++ b/net/dns/resolver/doh_test.go
@@ -15,10 +15,13 @@ import (
 
 var testDoH = flag.Bool("test-doh", false, "do real DoH tests against the network")
 
+const someDNSID = 123 // something non-zero as a test; in violation of spec's SHOULD of 0
+
 func someDNSQuestion(t testing.TB) []byte {
        b := dnsmessage.NewBuilder(nil, dnsmessage.Header{
                OpCode:           0, // query
                RecursionDesired: true,
+               ID:               someDNSID,
        })
        b.StartQuestions() // err
        b.Question(dnsmessage.Question{
@@ -56,9 +59,13 @@ func TestDoH(t *testing.T) {
                        c.Transport.(*http.Transport).CloseIdleConnections()
 
                        var p dnsmessage.Parser
-                       if _, err := p.Start(res); err != nil {
+                       h, err := p.Start(res)
+                       if err != nil {
                                t.Fatal(err)
                        }
+                       if h.ID != someDNSID {
+                               t.Errorf("response DNS ID = %v; want %v", h.ID, someDNSID)
+                       }
 
                        p.SkipAllQuestions()
                        aa, err := p.AllAnswers()
```

Reply:
Okay, fair. It's in a section talking about improving HTTP cache behavior, which POST obliterates anyway so that's probably fine.